File: customer_portal/custom_api/statement.py
import frappe
from frappe.utils import flt

@frappe.whitelist()
def get_customers_statement(from_date=None, to_date=None):
    # Check if `from_date` and `to_date` are provided
    if not from_date or not to_date:
        return {"error": "Both 'from_date' and 'to_date' are required."}

    # Get the currently logged-in user
    customer_email = frappe.session.user

    # Fetch customer record using the logged-in user's email
    customer = frappe.get_all('Customer', filters={'custom_link_user_email': customer_email}, fields=['name', 'customer_name'])
    
    # Check if a customer was found for the email
    if not customer:
        return {"error": "No customer found for the logged-in user."}
    
    customer_name = customer[0].name  # Get customer name for queries
    
    # Initialize variables
    sales_invoice_data = []
    total_paid_amount = 0
    grand_total_amount = 0
    running_balance = 0
    balance_brought_forward = 0

    # Calculate Balance Brought Forward
    previous_balance = frappe.db.sql("""
        SELECT 
            SUM(gle.debit - gle.credit) AS balance
        FROM 
            `tabGL Entry` gle
        WHERE 
            gle.party_type = 'Customer'
            AND gle.party = %s
            AND gle.posting_date < %s
            AND gle.docstatus = 1
    """, (customer_name, from_date), as_dict=True)
    
    if previous_balance and previous_balance[0].balance:
        balance_brought_forward = flt(previous_balance[0].balance)
    
    running_balance = balance_brought_forward

    # Step 1: Fetch Sales Invoices for the specified customer and date range, excluding canceled invoices
    invoices = frappe.db.sql("""
        SELECT 
            si.name AS invoice_name,
            si.posting_date,
            si.cost_center, 
            sii.item_code, 
            sii.qty, 
            sii.rate, 
            sii.amount
        FROM 
            `tabSales Invoice` si
        JOIN 
            `tabSales Invoice Item` sii ON sii.parent = si.name
        WHERE 
            si.customer = %s 
            AND si.posting_date BETWEEN %s AND %s 
            AND si.docstatus = 1
    """, (customer_name, from_date, to_date), as_dict=True)

    for invoice in invoices:
        total_amount = flt(invoice.amount)
        grand_total_amount += total_amount
        running_balance += total_amount
        
        sales_invoice_data.append({
            "invoice_name": invoice.invoice_name,
            "cost_center": invoice.cost_center,
            "posting_date": invoice.posting_date,
            "item_code": invoice.item_code,
            "qty": flt(invoice.qty),
            "rate": flt(invoice.rate),
            "amount": total_amount,
            "running_balance": running_balance
        })

    # Step 2: Fetch Payment Entries within the date range, excluding canceled payments
    payments = frappe.db.sql("""
        SELECT 
            pe.name AS payment_entry_name, 
            pe.posting_date,
            pe.cost_center, 
            pe.paid_amount
        FROM 
            `tabPayment Entry` pe
        WHERE 
            pe.party_type = 'Customer'
            AND pe.party = %s
            AND pe.posting_date BETWEEN %s AND %s
            AND pe.docstatus = 1
    """, (customer_name, from_date, to_date), as_dict=True)

    filtered_payments = []
    for payment in payments:
        payment_amount = flt(payment.paid_amount)
        total_paid_amount += payment_amount
        running_balance -= payment_amount
        
        filtered_payments.append({
            "payment_entry_name": payment.payment_entry_name,
            "cost_center": payment.cost_center,
            "posting_date": payment.posting_date,
            "paid_amount": payment_amount,
            "running_balance": running_balance
        })

    # Step 3: Fetch Journal Entry GL Entries for the customer, excluding canceled entries
    gl_entries = frappe.db.sql("""
        SELECT
            gle.name AS gl_entry_name,
            gle.posting_date,
            gle.cost_center,
            gle.debit,
            gle.voucher_no,
            gle.credit,
            gle.remarks
        FROM
            `tabGL Entry` gle
        WHERE
            gle.party_type = 'Customer'
            AND gle.party = %s
            AND gle.voucher_type = 'Journal Entry'
            AND gle.posting_date BETWEEN %s AND %s
            AND gle.docstatus = 1
    """, (customer_name, from_date, to_date), as_dict=True)

    filtered_gl_entries = []
    for gl_entry in gl_entries:
        debit = flt(gl_entry.debit)
        credit = flt(gl_entry.credit)
        running_balance += debit - credit
        total_paid_amount += credit
        
        filtered_gl_entries.append({
            "gl_entry_name": gl_entry.gl_entry_name,
            "posting_date": gl_entry.posting_date,
            "cost_center": gl_entry.cost_center,
            "voucher_no": gl_entry.voucher_no,
            "debit": debit,
            "credit": credit,
            "remarks": gl_entry.remarks,
            "running_balance": running_balance
        })

    # Calculate outstanding amount
    outstanding_amount = grand_total_amount - total_paid_amount

    return {
        "customer": customer_name,
        "balance_brought_forward": balance_brought_forward,
        "grand_total_amount": grand_total_amount,
        "total_paid_amount": total_paid_amount,
        "outstanding_amount": outstanding_amount,
        "sales_invoice_data": sales_invoice_data,
        "payments": filtered_payments,
        "gl_entries": filtered_gl_entries,
        "Customer_name": gl_entries
    }

File: customer_portal/www/client_portal/statement/statement.js
function fetchStatementData() {
    const fromDate = document.getElementById('from_date').value;
    const toDate = document.getElementById('to_date').value;

    // Constructing the API URL with query parameters for date filters
    const apiUrl = `/api/method/customer_portal.custom_api.statement.get_customers_statement?from_date=${fromDate}&to_date=${toDate}`;

    // Fetching the data from the API
    fetch(apiUrl)
        .then(response => response.json())
        .then(data => {
            const statement = data.message;

            // Populate statement summary
            document.getElementById('customer_name').innerText = statement.customer || 'N/A';
            document.getElementById('balance_brought_forward').innerText = formatCurrency(statement.balance_brought_forward) || '₦0';
            document.getElementById('grand_total_amount').innerText = formatCurrency(statement.grand_total_amount) || '₦0';
            document.getElementById('total_paid_amount').innerText = formatCurrency(statement.total_paid_amount) || '₦0';
            document.getElementById('outstanding_amount').innerText = formatCurrency(statement.outstanding_amount) || '₦0';

            const tableBody = document.getElementById('statement_data');
            tableBody.innerHTML = ''; // Clear previous rows

            // Helper function to create rows for different data types
            function createRow(type, postingDate, amount, paymentAmount, runningBalance, details = '') {
                const row = document.createElement('tr');
                row.innerHTML = `
                    <td>${type}</td>
                    <td>${postingDate}</td>
                    <td>${formatCurrency(amount)}</td>
                    <td>${formatCurrency(paymentAmount)}</td> <!-- Payment Amount Column -->
                    <td>${formatCurrency(runningBalance)}</td>
                    <td>${details}</td>
                `;
                tableBody.appendChild(row);
            }

            // Track the running balance starting with balance brought forward
            let currentRunningBalance = parseFloat(statement.balance_brought_forward) || 0;

            // Process Sales Invoice Data
            if (statement.sales_invoice_data && statement.sales_invoice_data.length > 0) {
                statement.sales_invoice_data.forEach(invoice => {
                    currentRunningBalance += parseFloat(invoice.amount);
                    createRow(
                        'Invoice',
                        invoice.posting_date,
                        invoice.amount,
                        0, // No payment amount for invoice
                        currentRunningBalance.toFixed(2),
                        `Invoice Name: ${invoice.invoice_name}, Item: ${invoice.item_code}`
                    );
                });
            }

            // Process Payment Data
            if (statement.payments && statement.payments.length > 0) {
                statement.payments.forEach(payment => {
                    currentRunningBalance -= parseFloat(payment.paid_amount);
                    createRow(
                        'Payment',
                        payment.posting_date,
                        0, // No amount for payment in this column
                        payment.paid_amount, // Payment amount in the separate column
                        currentRunningBalance.toFixed(2),
                        `Payment Entry: ${payment.payment_entry_name}`
                    );
                });
            }

            // Process GL Entries Data
            if (statement.gl_entries && statement.gl_entries.length > 0) {
                statement.gl_entries.forEach(glEntry => {
                    currentRunningBalance -= parseFloat(glEntry.debit); // Debit reduces balance
                    createRow(
                        'GL Entry',
                        glEntry.posting_date,
                        glEntry.debit,
                        0, // No payment amount for GL Entry
                        currentRunningBalance.toFixed(2),
                        `Voucher No: ${glEntry.voucher_no}, Cost Center: ${glEntry.cost_center}`
                    );
                });
            }
        })
        .catch(error => console.error('Error fetching data:', error));
}

// Function to format currency as Naira (₦)
function formatCurrency(amount) {
    // Ensure the amount is a valid number
    if (isNaN(amount)) {
        return '₦0.00';
    }

    // Convert the amount to a number and format it as Naira currency
    const number = parseFloat(amount);
    return new Intl.NumberFormat('en-NG', {
        style: 'currency',
        currency: 'NGN',
    }).format(number);
}

function downloadStatement() {
    // Extract the content to be downloaded
    var contentToDownload = `
        <div class="content">
            <div class="company-header">
                <img src="${document.getElementById('logo').src}" alt="Company Logo" class="company-logo">
                <div class="company-info">
                    <h2>METRO AND CASTLE LIMITED</h2>
                    <p>METRO MALL, ATLANTIC LAYOUT ESTATES,</p>
            <p>LEKKI GARDEN PHASE 4 ROAD, GENERAL PAINT B/STOP,</p>
            <p>LEKKI EXPRESSWAY, LAGOS.</p>
            <p>PHONE: 09154793665, 08097814209</p>
            <p>Info@metroandcastle.com www.metroandcastle.com</p></center>
                </div>
            </div>
            <h1>Customer Statement</h1>
            <div class="balances">
                <div class="balance-item">
                    <p><strong>Customer:</strong> <span id="customer_name">${document.getElementById('customer_name').textContent}</span></p>
                    <p><strong>Balance Brought Forward:</strong> <span id="balance_brought_forward">${document.getElementById('balance_brought_forward').textContent}</span></p>
                </div>
                <div class="balance-item">
                    <p><strong>Grand Total Amount:</strong> <span id="grand_total_amount">${document.getElementById('grand_total_amount').textContent}</span></p>
                    <p><strong>Total Paid Amount:</strong> <span id="total_paid_amount">${document.getElementById('total_paid_amount').textContent}</span></p>
                    <p><strong>Outstanding Amount:</strong> <span id="outstanding_amount">${document.getElementById('outstanding_amount').textContent}</span></p>
                </div>
            </div>
            <table class="statement-table">
                <thead>
                    <tr>
                        <th>Type</th>
                        <th>Posting Date</th>
                        <th>Amount</th>
                        <th>Payment Amount</th>
                        <th>Running Balance</th>
                        <th>Details</th>
                    </tr>
                </thead>
                <tbody>
                    ${document.getElementById('statement_data').innerHTML}
                </tbody>
            </table>
        </div>
        <style>
            body {
                font-family: 'Arial', sans-serif;
                background-color: #f4f4f9;
                margin: 0;
                padding: 0;
                color: #333;
            }

            .content {
                background-color: #ffffff;
                padding: 30px;
                border-radius: 8px;
                max-width: 800px;
                margin: 0 auto;
                box-shadow: 0 4px 10px rgba(0, 0, 0, 0.1);
            }

            .company-header {
                display: flex;
                align-items: center;
                justify-content: space-between;
                margin-bottom: 20px;
                border-bottom: 2px solid #f1f1f1;
                padding-bottom: 10px;
            }

            .company-logo {
                width: 300px; /* Adjust size of the logo */
                height: auto;
            }

            .company-info h2 {
                color: #4A90E2;
                margin: 0;
                font-size: 24px;
            }

            .company-info p {
                font-size: 14px;
                color: #555;
                margin: 0;
            }

            h1 {
                color: #4A90E2;
                font-size: 24px;
                text-align: center;
                margin-bottom: 20px;
            }

            .balances {
                display: flex;
                justify-content: space-between;
                margin-bottom: 30px;
                padding: 10px 0;
                border-bottom: 2px solid #f1f1f1;
            }

            .balance-item {
                width: 48%;
                display: flex;
                flex-direction: column;
            }

            .balances p {
                font-size: 14px;
                margin: 8px 0;
                color: #555;
            }

            .balances strong {
                color: #4A90E2;
            }

            table {
                width: 100%;
                border-collapse: collapse;
                margin-top: 20px;
            }

            table th,
            table td {
                padding: 12px;
                text-align: left;
                border: 1px solid #f1f1f1;
                font-size: 14px;
            }

            table th {
                background-color: #4A90E2;
                color: white;
            }

            table td {
                background-color: #f9f9f9;
            }

            table tr:nth-child(even) td {
                background-color: #f1f1f1;
            }

            table tr:hover td {
                background-color: #e1e1e1;
            }

            @media print {
                body {
                    background-color: #fff;
                }

                .content {
                    box-shadow: none;
                    margin: 0;
                    padding: 20px;
                    width: 100%;
                }

                table {
                    border: 1px solid #ddd;
                }

                table th {
                    background-color: #333;
                    color: #fff;
                }
            }
        </style>
    `;

    // Create a Blob from the content
    var blob = new Blob([contentToDownload], { type: 'text/html' });

    // Create a link element for downloading the content
    var link = document.createElement('a');
    link.href = URL.createObjectURL(blob);
    link.download = 'customer_statement.html'; // Name of the downloaded file

    // Programmatically click the link to trigger the download
    link.click();
}



// print PDF SECTION


